### ISSUE TITLE
EPU3 Media Overlays: Next + Previous granularity

### ISSUE THREAD

**Issue by ManfredMuchenberger** [edrlab/thorium-reader#1070]
IThorium v1.3.1-alpha.1.40
Windows 10

In a books with media overlays and word level synchronization.
When I press the "Next Readaloud" or the "Previous Readaloud" button the playback jumps to the next or previous word.
As a user I would expect a jump to the next phrase or maybe better a user setting where I can set the "unit" of the jump (e.g. Phrase, Sentence, Chapter, Time-Jump).
A one-word-jump is not usefull.

**Comment by danielweck** [comment]
Interesting feature suggestion.

So, how can a reading system know whether a SMIL-synchronized unit of text/audio is a "word", versus a "sentence", versus a "paragraph"? (etc.) In other words, is there a vocabulary (e.g. `epub:type` attribute values) that can be relied upon to control "navigation granularity", distinctly from the "highlighting granularity"?

Note that in Readium1 there was an experimental feature which allowed users to control such multiple granularity, but this was not based on an official standard, it was just a prototype. Navigation granularity could be switched on the fly.

**Comment by ManfredMuchenberger** [comment]
There is already a lot of tags to process in a book with word-level synchronization. So I don't know if its a good idea to add more attributes. 
Probably most reading systems are trying to analyse the text somehow.
I am not a HTML expert but trying to describe it anyway. 

Ideal would be the possibilty to choose from these navigation granularitys:
1. only headings with level <h1>
2. all headings
3. <p> for paragraph should be easy
4. for sentences you would either have to look for the "." or just punctuation marks outside a <span>. e.g. <span id="w00392">heulte</span>. <span id="w00393">Genau</span>. 
5. a word is the smalest unit and contains no blanks <span id="w00253">ihr</span>
6. most reading systems also have "time jumps" as a "unit". They probably calculated them by adding the "ms" in the SMIL <audio src="../audio/section-0003.mp3" clipBegin="5255ms" clipEnd="6385ms"/>

I will add some screenshots from EasyReader and VoiceDreamReader apps.

![Screenshot_20200529-080547_Reader](https://user-images.githubusercontent.com/11869355/83227125-b5235200-a183-11ea-8ae2-86e4a248d331.jpg)
![Screenshot_20200529-080442_EasyReader](https://user-images.githubusercontent.com/11869355/83227130-b94f6f80-a183-11ea-8d4e-0e834ab8add3.jpg)

**Comment by danielweck** [comment]
I've unzipped one of the EPUBs you provided (thank you!) and neither the SMIL XML nor the (X)HTML markup contain semantic information (such as `epub:type` or `role` attributes) from which to reliably extract the notion of "word" versus "sentence" vs. other higher granularity levels (for the purpose of synchronization/highlight, and-or for the purpose of previous/next linear navigation units).

Consequently, this must be inferred from the HTML markup based on some ad-hoc heuristic, for example to assume that `span` elements are "words" (which is far from always being the case), and that the parent `p` (if there is such parent) is a "sentence" (again, not necessarily the case).

Frankly, this seems very brittle to me, in that what will work with one set of publications will break in another. In the absence of clear specification for this feature, I am concerned that our best effort would be a vague approximation of the desired UX, overall.

Note that the DAISY concept of "escapability" is a good fit for this feature, but this isn't usually based on guesswork, instead this normally relies on robust semantics.

A more technical note: strictly-speaking, the EPUB3 Media Overlays playback engine follows the SMIL timeline "to the letter", HTML highlighting is merely a side effect of the text/audio synchronization process. In other words, at authoring time the semantic inflection normally found in HTML documents (in layman terms: `epub:type` attributes and associated vocabulary) is mirrored in the SMIL XML whenever necessary (for example, to reflect "bookmatter", "section", etc.). As a result, at runtime / playback time, the engine picks up semantic information directly from the SMIL which is the single source of truth. The aforementioned techniques to attempt to infer structural semantics from HTML markup would require introspecting the DOM either ahead of time (which would introduce significant computational costs) or on the fly during playback (which would be a departure from the standard processing model reliant on the SMIL(s)).

With that in mind, I am happy to hear that other reading systems seem to be implementing support for multiple navigation granularities despite the lack of interoperable specification, and I would love to hear more about their approach to solving this problem.

Ultimately, I totally agree that word-by-word highlighting could/should be conceptually separated from the higher-level previous/next linear navigation model. I hope we can find a workable solution.

**Comment by danielweck** [comment]
PS: thank you very much for the screenshots! Could you please translate to English? (I just do not want to miss important clues from their UI)

**Comment by ManfredMuchenberger** [comment]
Just added the translation below the screenshots.
I have also added a remark above the screenshots to make that clear: "They so far don't support Media Overlays. So this is the navigation in a text-only EPUB using the TTS that is built in the apps."

**Comment by danielweck** [comment]
Ah, TTS readaloud is completely different than EPUB3 Media Overlays, in that the UX in the former case is dictated by the reading system / at runtime, whereas in the latter case the UX is in large part informed by content creators / at authoring time.

Thorium also supports TTS readaloud, and we are working on improvements that will be released soon. We already support word-by-word highlighting with sentence-level previous/next navigation. You may test this functionality with any non-MediaOverlays EPUBs. We also welcome feedback for the TTS feature :)

**Comment by danielweck** [comment]
So, if I understand correctly, this is not an issue with Thorium, right?

To summarize:

1) TTS readaloud (live synthetic speech): Thorium supports word-level highlighting and sentence-level navigation (i.e. the granularity of individual utteration, which defines the atomic unit of previous/next linear navigation). There is an in-document view (which is the default mode), as well as a "clean" view which removes most formatting except for headings and images with alt attributes (this mode is optional, users must explicitly activate it).
2) EPUB3 Media Overlays readaloud (pre-recorded audio files): Thorium follows the granularity authored in the SMIL XML, which can be word-level, sentence-level, paragraph-level, etc. This is the atomic unit for both highlighting and previous/next linear navigation.

I am closing this issue now (bug triage), but please feel free to respond and I will be happy to re-open, to continue the discussion :)
